include/contraction/contractionGraph.hpp

```diff
@@ -47,60 +47,112 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+            ++out
+        ) 
+        adjacent_vertices += this->adjacent(v, *out);
+
+        for (
+            boost::tie(in, in_end) = in_edges(v, this->graph);
+            in != in_end;
+            ++in
+        )
+        adjacent_vertices += this->adjacent(v, *in);
+        
         return adjacent_vertices;
     }
 
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    std::vector<E> get_shortcuts() {
+        pgrouting::Identifiers<E> eids;
+        E_i e, e_end;
+        for (boost::tie(e, e_end) = edges(this->graph); e != e_end; ++e) {
+            if (this->graph[*e].id < 0) {
+                eids += *e;
+                pgassert(!(this->graph[*e]).get_contracted_vertices().empty());
+            } else {
+                pgassert((this->graph[*e]).get_contracted_vertices().empty());
+            }
+        }
+        std::vector<E> o_eids(eids.begin(), eids.end());
+        std::sort(
+            o_eids.begin(), 
+            o_eids.end(),
+            [&](E lhs, E rhs) {
+                return -1*(this->graph)[lhs].id < -1*(this->graph)[rhs].id;
+            }
+        );
+        return o_eids;
+    }
+
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    Identifiers<int64_t> get_modified_vertices() {
+        Identifiers<int64_t> vids;
+        for (auto v : boost::make_iterator_range(boost::vertices(this->graph))) {
+            if ((this->graph[v]).has_contracted_vertices()) {
+                vids += (this->graph[v]).get_id();
+            }
+        }
+        return vids;
+    }
 
-    /*! @brief get the edge with minimum cost between two vertices
-    @param [in] u vertex_descriptor of source vertex
-    @param [in] v vertex_descriptor of target vertex
-    @return E: The edge descriptor of the edge with minimum cost
+    /*!
+        @brief get the edge with minimum cost between two vertices
+        @param [in] u vertex_descriptor of source vertex
+        @param [in] v vertex_descriptor of target vertex
+        @return E: The edge descriptor of the edge with minimum cost
     */
-    std::tuple<double, Identifiers<int64_t>, bool>
-    get_min_cost_edge(V u, V v) {
+    std::tuple<double, Identifiers<int64_t>, bool> get_min_cost_edge(V u, V v) {
         E min_edge;
         Identifiers<int64_t> contracted_vertices;
         double min_cost = (std::numeric_limits<double>::max)();
         bool found = false;
 
         if (this->is_directed()) {
             BGL_FORALL_OUTEDGES_T(u, e, this->graph, G) {
-                if (this->target(e) == v) {
-                    contracted_vertices += this->graph[e].contracted_vertices();
-                    if (this->graph[e].cost < min_cost) {
-                        min_cost = this->graph[e].cost;
+                if (target(e, this->graph) == v) {
```

Comment:
This doesn't seem like the previous `target` call. Could you elaborate on which `target` do you refer there?

Reply:
Hum, it seems the code changed since last time, I was puzzled as `target` was called with two different manners. But let's say it is OK, and I got it for the boost fact.